src/rules/strictStringExpressionsRule.ts

```diff
@@ -0,0 +1,159 @@
+/**
+ * @license
+ * Copyright 2019 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isTypeFlagSet, isUnionType } from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+const OPTION_ALLOW_EMPTY_TYPES = "allow-empty-types";
+
+interface Options {
+    [OPTION_ALLOW_EMPTY_TYPES]?: boolean;
+}
+
+export class Rule extends Lint.Rules.TypedRule {
+    public static CONVERSION_REQUIRED = "Explicit conversion to string type required";
+
+    public static metadata: Lint.IRuleMetadata = {
+        description: "Disable implicit toString() calls",
+        descriptionDetails: Lint.Utils.dedent`
+            Require explicit toString() call for variables used in strings. By default only strings are allowed.
+
+            The following nodes are checked:
+
+            * String literals ("foo" + bar)
+            * ES6 templates (\`foo \${bar}\`)`,
+        hasFix: true,
+        optionExamples: [
+            true,
+            [
+                true,
+                {
+                    [OPTION_ALLOW_EMPTY_TYPES]: true,
+                },
+            ],
+        ],
+        options: {
+            properties: {
+                [OPTION_ALLOW_EMPTY_TYPES]: {
+                    type: "boolean",
+                },
+            },
+            type: "object",
+        },
+        optionsDescription: Lint.Utils.dedent`
+                Following arguments may be optionally provided:
+                * \`${OPTION_ALLOW_EMPTY_TYPES}\` allows \`null\`, \`undefined\` and \`never\` to be passed into strings without explicit conversion`,
+        requiresTypeInfo: true,
+        ruleName: "strict-string-expressions",
+        type: "functionality",
+        typescriptOnly: true,
+    };
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        return this.applyWithFunction(
+            sourceFile,
+            walk,
+            this.getRuleOptions(),
+            program.getTypeChecker(),
+        );
+    }
+
+    private getRuleOptions(): Options {
+        if (this.ruleArguments[0] === undefined) {
+            return {
+                [OPTION_ALLOW_EMPTY_TYPES]: true,
```

Comment:
empty types are enabled by default

Reply:
I'm not sure about name though.. can `never` be considered as `empty`?

Reply:
Great semantic question. Technically `never` is a [https://en.wikipedia.org/wiki/Bottom_type](bottom type); however, `void`, `undefined`, etc. are separate types all on their own. `empty` seems like a good way to refer to all these.

Reply:
yeah, that was difficult to find common name for these